Add blog post for optimized unnest

yingsu00 opened the pull request.

yingsu00 commented:
yarn start result:

![Screen Shot 2020-07-20 at 3 00 57 AM](https://user-images.githubusercontent.com/33299678/87925743-90d24c00-ca35-11ea-8887-4d8df429d818.png)
![Screen Shot 2020-07-20 at 3 01 21 AM](https://user-images.githubusercontent.com/33299678/87925741-9039b580-ca35-11ea-9a72-8e898449da6d.png)
![Screen Shot 2020-07-20 at 3 01 35 AM](https://user-images.githubusercontent.com/33299678/87925739-9039b580-ca35-11ea-944a-5d1bd8e71f70.png)
![Screen Shot 2020-07-20 at 3 01 47 AM](https://user-images.githubusercontent.com/33299678/87925737-8fa11f00-ca35-11ea-91b3-b1a540a0e772.png)
![Screen Shot 2020-07-20 at 3 01 57 AM](https://user-images.githubusercontent.com/33299678/87925736-8fa11f00-ca35-11ea-8c54-c62893bf206f.png)
![Screen Shot 2020-07-20 at 3 02 10 AM](https://user-images.githubusercontent.com/33299678/87925735-8f088880-ca35-11ea-8f7d-868385409f41.png)
![Screen Shot 2020-07-20 at 3 02 20 AM](https://user-images.githubusercontent.com/33299678/87925732-8e6ff200-ca35-11ea-90fb-683fcc612551.png)
![Screen Shot 2020-07-20 at 3 02 30 AM](https://user-images.githubusercontent.com/33299678/87925730-8dd75b80-ca35-11ea-9aa5-1de12c81a089.png)
![Screen Shot 2020-07-20 at 3 02 37 AM](https://user-images.githubusercontent.com/33299678/87925725-8b750180-ca35-11ea-8d6e-3510d96e5a62.png)

aweisberg commented on website/blog/2020-07-17-unnest.md:
```diff
@@ -0,0 +1,219 @@
+
+To make the above loop possible, we pre-compute the cardinalities for each row for all unnest columns and use them to get the max cardinalities (we call it the maxEntries array) in the operator. This way the memory can be accessed sequentially and the computation can be reused. The size of these arrays should require very little memory overhead, since they’re only for top level rows and should al ...
+
+We then compute the batch size, ie. top level row count that can fit into the next output blocks. Right now we limit the nested row count for each batch 1000. The batch size is at least 1. This means the unnested row count for each block might be over 1000. For example, if a row contains 10000 array entries, then the unnested row count would be 10000. Although this is over the 1000 rows we still  ...
```
output blocks or output block? It's outputting one block at a time so probably singular?

"Right now we limit the nested row count for each batch to 1000 with a minimum batch size of 1" - added "to 1000" and combined the two sentences.

yingsu00 replied:
> output blocks or output block? It's outputting one block at a time so probably singular?
> 
It's blocks. We calculate batch size for all blocks. To avoid confusion, I changed it to `output page`.

aweisberg commented on website/blog/2020-07-17-unnest.md:
```diff
@@ -0,0 +1,219 @@
+
+There are several cases that we can tell no `NULL`s need to be inserted:
+
+1. When there is only one unnest column (except Array of Row case).
```
`ARRAY` of `ROW`s case

aweisberg commented on website/blog/2020-07-17-unnest.md:
```diff
@@ -0,0 +1,219 @@
+
+1. When there is only one unnest column (except Array of Row case).
+2. When there are multiple columns, but the cardinalities for them are the same for the current batch.
+3. For a single Array of Row unnest column case, if the row block doesn't contain any nulls.
```
`ARRAY` of `ROW`s

aweisberg replied:
Marked resolved, but it's still not in a code block

yingsu00 commented on website/blog/2020-07-17-unnest.md:
```diff
@@ -0,0 +1,219 @@
+
+Ying Su, Masha Basmanova, Orri Erling
+
+Unnest is a common operation in Facebook’s daily Presto workload. It converts an array, map or row into a flat relation. Its original implementation uses deep copy all the time and was very inefficient. In https://prestosql.io/blog/2019/08/23/unnest-operator-performance-enhancements.html, the author improved the Unnest operator by up to 10x in CPU and elapsed times by using `DictionaryBlock` when ...
```
@aweisberg Ariel, thank you so much for reviewing the blog!! I'm sorry I didn't notice you have reviewed it. I'm addressing your comments now. 

>In this post, we will refer to the original implementation as "baseline".

The "baseline" is actually the PrestoSQL's version. So I changed this sentence to
```
In this post, we will refer to the PrestoSQL implementation as "baseline".
```
Do you think referring to PrestoSQL is ok? Would it cause any problems?

yingsu00 commented on website/blog/2020-07-17-unnest.md:
```diff
@@ -0,0 +1,219 @@
+![Remote Exchange](/img/blog/2020-07-17-unnest.md/UnnestOneColumnWithoutOrdinality.png)
+![Remote Exchange](/img/blog/2020-07-17-unnest.md/UnnestOneColumnWithOrdinality.png)
+![Remote Exchange](/img/blog/2020-07-17-unnest.md/UnnestTwoColumns.png)
+![Remote Exchange](/img/blog/2020-07-17-unnest.md/UnnestRowType.png)
```
> These images are hard to read and can't be enlarged by clicking. It would also be good to link to the JMH benchmark so readers can understand what was measured.

@aweisberg did you mean the original benchmark results in its original format? Or a link to the benchmark source code?
I enlarged the fonts in these images. These images looked vague because I took screen shot on the whole page. The real blog should display them better. Also added `The benchmark can be found in presto-main/src/test/java/com/facebook/presto/operator/BenchmarkUnnestOperator.java.`

aweisberg replied:
Hey the images look a bit unclear on the actual page. I am running the web site locally with yarn start. I was thinking it would be nice if the image could be a link to a larger sharper version.

I was thinking link to the actual benchmark.

yingsu00 commented on website/blog/2020-07-17-unnest.md:
```diff
@@ -0,0 +1,219 @@
+![Remote Exchange](/img/blog/2020-07-17-unnest.md/UnnestRowType.png)
+
+The optimized Unnest implementation has been merged in Presto master, and was available since version 0.235. It is enabled by default and no configuration or session property is required.
+In Facebook production workload, we observed over 6x CPU gain on this operator. The following chart shows the operator's CPU percentage was reduced from about 2% to 0.1% out of all operators after the roll-out happened on May 28:
```
Thanks Ariel. I reordered the images, and added this:
```
About 25% operators show over 5x CPU reduction, and some of them are even over 20x reduction.
```

yingsu00 commented on website/blog/2020-07-17-unnest.md:
````diff
@@ -0,0 +1,219 @@
+| SanFranciscoZoo | cat    |
+```
+
+In Presto engine, the relational data is represented as `Page` internally, which in turn is composed of `Block`'s, one for each column. For the replicated column, the Unnest operator needs to create a new `Block` from the original `Block` with the data being replicated within each original row. The baseline implementation achieved this by creating a `DictionaryBlock` on top of the original `Block ...
````
> What is a replicated column, can you explain that?

Sure. I changed the above sentence to 
```
In this example, the `zoo` column, which is a `VARCHAR` column, is being replicated for the same row and thus being called "replicated column"; and the `animals` column, which is an `ARRAY` column, is unnested into a `VARCHAR` column, named as `animal` and is a "unnest column":
```

> New paragraph at "The baseline" ?

"The baseline" needs to stay with the explanation of how replicated column was handled in the last sentence. And then we talk about the unnest columns. Is it too long to read?

yingsu00 commented:
@aweisberg Hi Ariel, thank you very much for your review! I have addressed all your comments. Would you mind giving it another look?

aweisberg commented on website/blog/2020-07-17-unnest.md:
```diff
@@ -0,0 +1,221 @@
+
+Ying Su, Masha Basmanova, Orri Erling
+
+Unnest is a common operation in Facebook’s daily Presto workload. It converts an `ARRAY`, `MAP`, or `ROW` into a flat relation. Its original implementation uses deep copy all the time and was very inefficient. In https://prestosql.io/blog/2019/08/23/unnest-operator-performance-enhancements.html, the author improved the Unnest operator by up to 10x in CPU and elapsed times by using `DictionaryBloc ...
```
Minor improvement, but can you make that not a raw link but instead a text link to it?

aweisberg commented on website/blog/2020-07-17-unnest.md:
```diff
@@ -0,0 +1,221 @@
+![Remote Exchange](/img/blog/2020-07-17-unnest.md/UnnestTwoColumns.png)
+![Remote Exchange](/img/blog/2020-07-17-unnest.md/UnnestRowType.png)
+
+The optimized Unnest implementation is available in Presto 0.235 and onward and is enabled by default. The benchmark can be found in `presto-main/src/test/java/com/facebook/presto/operator/BenchmarkUnnestOperator.java`.
```
Can you turn this into a link to the GH file location (freeze at at the commit that this blog post was written on) and just call it "BenchmarkUnnestOperator.java"? That will make it render nicer.

aweisberg commented on website/blog/2020-07-17-unnest.md:
```diff
@@ -0,0 +1,221 @@
+
+The optimized Unnest implementation is available in Presto 0.235 and onward and is enabled by default. The benchmark can be found in `presto-main/src/test/java/com/facebook/presto/operator/BenchmarkUnnestOperator.java`.
+
+In Facebook production workload, we observed over 6x CPU reduction on this operator. The following chart shows the pariwise comparison for all UnnestOperators in one of our test runs. Every dot below the diagonal is a win. About 25% operators show over 5x CPU reduction, and some of them have even over 20x reduction.
```
over 6x -> over a 6x
Maybe:
UnnestOperators -> unnest operators 

About 25% operators show over -> About 25% of operators show over a

have even over 20x -> have even over a 20x

aweisberg commented on website/blog/2020-07-17-unnest.md:
```diff
@@ -0,0 +1,221 @@
+
+![Remote Exchange](/img/blog/2020-07-17-unnest.md/UnnestCPUPairWiseComparison.png)
+
+The histogram of the after vs. before CPU ratio shows most operators has a ratio less than 1, meaning most of them were more efficient. 
```
has a ratio -> have a ratio

aweisberg commented on website/blog/2020-07-17-unnest.md:
```diff
@@ -0,0 +1,221 @@
+
+## What is the Unnest Operator?
+
+When users have data structured as `ARRAY`, `MAP` or `ROW` they sometimes need to flatten them so that the nested structure can be regarded as top level citizen and sent to downstream operators for easier arithmetic or aggregation processing. An example query is as follows:
```
, after `MAP`

aweisberg commented on website/blog/2020-07-17-unnest.md:
````diff
@@ -0,0 +1,221 @@
+| SanFranciscoZoo | cat    | Danny    |
+```
+
+Note that a `NULL` is inserted in the last column on the third row. In this case, can we still create a `DictionaryBlock` on the original block? The original nested block does not have a `NULL`! So it’s impossible to find an index that points to a `NULL` value and build a `DictionaryBlock` easily. The baseline implementation just switches back to deep copying for this case. In the sections below, ...
````
These additional explanations are excellent, thanks!

aweisberg commented on website/blog/2020-07-17-unnest.md:
```diff
@@ -0,0 +1,221 @@
+The baseline implementation did a number of things well. It avoided deep copy by using DictionaryBlock which improved the original improvement by an order of magnitude for cases without NULL insertions. Can we improve it additionally? The answer is yes.
+
+### Process data column-by-column, not row-by-row
+If you read our blog post https://prestodb.io/blog/2019/12/20/repartition, you may remember the 5 design choices we picked for repartitioning. The first was “Process data column by column, not row by row”. The same thing applies to `UnnestOperator` as well. The baseline implementation constructs the output blocks in the row-by-row manner:
```
Can you convert the link to be the name of the blog post instead of a raw URL?

yingsu00 commented:
@aweisberg Ariel, thank you so much for such a careful review! I addressed all your comments and renamed it from `2020-07-17-unnest.md` to `2020-08-20-unnest.md`. Do you mind review it again? Thank you!

aweisberg commented on website/blog/2020-08-20-unnest.md:
```diff
@@ -0,0 +1,221 @@
+
+The optimized Unnest implementation is available in Presto 0.235 and onward and is enabled by default. The JMH benchmark was also expanded to have better coverage and can be found in [BenchmarkUnnestOperator.java](https://github.com/prestodb/presto/blob/29acdbaff128fff2de185c28fb26031be1d88019/presto-main/src/test/java/com/facebook/presto/operator/BenchmarkUnnestOperator.java).
+
+In Facebook production workload, we observed over a 6x CPU reduction on this operator. The following chart shows the pariwise comparison for all unnest operatrs in one of our test runs. Every dot below the diagonal is a win. About 25% of operators show over a 5x CPU reduction, and some of them have even over a 20x reduction.
```
```suggestion
In Facebook's production workload, we observed over a 6x CPU reduction on this operator. The following chart shows the pairwise comparison for all unnest operators in one of our test runs. Every dot below the diagonal is a win. About 25% of operators show over a 5x CPU reduction, and some of them have even over a 20x reduction.
```

yingsu00 replied:
@aweisberg Ariel thank you so much! I have corrected these typos and moved the charts to `/img/blog/2020-08-20-unnest.md`. Will you be able to merge it?

wenleix commented:
merged #97. thanks for the contribution!

yingsu00 commented:
> merged #97. thanks for the contribution!

Thank you Wenlei! @wenleix
